### PR TITLE
[SYCL][COMPAT] Support for experimental masked shifts

### PR DESCRIPTION
**Pull request by Alcpz**
This PR adds an experimental approach to support `select_from_sub_group`,
`shift_sub_group_left`, `shift_sub_group_right`, and `permute_sub_group_by_xor` functions  on SPIRV devices.
 
As these are based on still to be supported intrinsics, tests have been left out.

### CODE REVIEW

**Comment by JackAKirk** [comment]
These can be immediately implemented using https://github.com/oneapi-src/SYCLomatic/pull/1342, and the macros used there have already been migrated to intel/llvm here:  https://github.com/intel/llvm/pull/13230
We could sync merging these PRs to sort this? : either get wait for that one to be merged and add the cuda impls, or get this one merged first and I can add the impls on the other one.

There's a few other things this compat stuff raises. Firstly it seems to be a mix of "compat" for std:: and "compat" for cuda? Is this fair to say?
For cuda stuff this creates a dilemma because we already have similarly migrated cuda/oneapi extensions that use a different namespace: see the functions implemented in `sycl::ext::oneapi::experimental::cuda`: see implementations and PRs from https://github.com/intel/llvm/tree/sycl/sycl/include/sycl/ext/oneapi/experimental/cuda. Should we first decide on a single namespace to use for such migrations? I don't think it is a good idea if we create a zoo of different namespaces for cuda translated functionality when we could just have one?

**Comment by Alcpz** [comment]
These changes are not up to date with the latest SYCLomatic version, as we had a set snapshot to port a fixed set of functionalities. These will need to be updated once more before finally removing all the dependencies from the headers you've pointed in that SYCLomatic link. 

You are right, we should not have two implementations and avoid the 'zoo' of namespaces. So any feature/API call that is achievable through SYCL should be wrapped in SYCLcompat instead of having its own separate implementation. 

Once this gets merged I think it's sensible to go through the implementation you have in `sycl::ext::oneapi::experimental::cuda`

**Review by AidanBeltonS — APPROVED** [review]
The code itself seems fine. I do see @JackAKirk 's point, ideally we should not have multiple implementations. I think it is fine for now as SYCLomatic will try to map to these functions, but should be merged with the other approach sooner than later.

**Comment by Alcpz** [comment]
Thanks for the input. Since it's already added as a TODO, I will ping for merging.

@intel/llvm-gatekeepers This PR is ready for merging. Thanks

**Pull request merged**
